e2etester/createtestworkflows.go

```diff
@@ -0,0 +1,243 @@
+	if ierr != nil {
+		return ierr.(error)
+	}
+	return nil
+}
+
+func createTestWorkflows(tests interface{}, images []string, regex *regexp.Regexp) (map[string]*daisy.Workflow, error) {
+	// Create base test workflows.
+	workflows := make(map[string]*daisy.Workflow)
+	typ := reflect.TypeOf(tests)
+	for i := 0; i < typ.NumMethod(); i++ {
+		m := typ.Method(i)
+		if regex != nil && !regex.MatchString(m.Name) {
+			continue
+		}
+		for _, image := range images {
+			wf, err := createVMWorkflow(m.Name, image)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create workflow for %q", m.Name)
+			}
+			nameAndImage := fmt.Sprintf("%s [%s]", m.Name, image)
+			workflows[nameAndImage] = wf
+		}
+	}
+
+	// Include dependency tests as workflows in their dependents.
+	for dependentTest, dependeeTest := range dependencies {
+		for _, image := range images {
+			dependentName := fmt.Sprintf("%s [%s]", dependentTest, image)
+			dependentwf, ok := workflows[dependentName]
+			if !ok {
+				return nil, fmt.Errorf("%q in dependency map, but not found", dependentName)
+			}
+			dependeeName := fmt.Sprintf("%s [%s]", dependeeTest, image)
+			dependeewf, ok := workflows[dependeeName]
+			if !ok {
+				return nil, fmt.Errorf("%q in dependency map, but not found", dependeeName)
+			}
+
+			// Make dependee instance name available to the dependent.
+			dependentCreateStep := dependentwf.Steps[dependentTest]                                     // create instance step is named after test.
+			dependentInstance := (*dependentCreateStep.CreateInstances)[0]                              // pull out the instance config.
+			dependentInstance.Metadata[dependeeTest] = fmt.Sprintf("test-instance-%s", dependeewf.ID()) // add an entry to dependent metadata.
+
+			step, err := dependentwf.NewStep("include")
+			if err != nil {
+				return nil, fmt.Errorf("Failed to create workflow step: %v", err)
+			}
+			step.SubWorkflow = &daisy.SubWorkflow{Workflow: dependeewf}
+			delete(workflows, dependeeName) // dependee is no longer a top-level WF.
+
+			// dependent create-instance step run after the dependee workflow.
+			dependentwf.AddDependency(dependentwf.Steps[dependentTest], dependentwf.Steps["include"])
+		}
+	}
+
+	// Add wait signals to top-level WFs.
+	for _, wf := range workflows {
+		waitStepName := fmt.Sprintf("wait-instance-%s", wf.Name)
+		step, err := wf.NewStep(waitStepName)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to create workflow step: %v", err)
+		}
+		instanceSignal := &daisy.InstanceSignal{
+			Name: wf.Name,
+			SerialOutput: &daisy.SerialOutput{
+				Port:         1,
+				SuccessMatch: "TEST-SUCCESS",
+				FailureMatch: daisy.FailureMatches{"TEST-FAILURE"},
+			},
+		}
+		step.WaitForInstancesSignal = &daisy.WaitForInstancesSignal{instanceSignal}
+		step.Timeout = "2m"
+		wf.AddDependency(wf.Steps[waitStepName], wf.Steps[wf.Name])
+	}
+
+	return workflows, nil
+}
+
+func createVMWorkflow(name, image string) (*daisy.Workflow, error) {
```

Comment:
What's the difference between Test and VM workflows?

Reply:
VM workflow is specific to one VM - creates a disk, creates an instance. a test workflow is the overall test, which may have multiple VM workflows embedded.